homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,192 @@
+
+import contextlib
+import logging
+from typing import Any
+
+from pysonos.core import SoCo
+from pysonos.events_base import Event as SonosEvent
+from pysonos.exceptions import SoCoException
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.const import DEVICE_CLASS_BATTERY, PERCENTAGE, STATE_UNKNOWN
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.icon import icon_for_battery_level
+
+from . import SonosData
+from .const import (
+    BATTERY_SCAN_INTERVAL,
+    DATA_SONOS,
+    SONOS_DISCOVERY_UPDATE,
+    SONOS_ENTITY_CREATED,
+    SONOS_PROPERTIES_UPDATE,
+)
+from .entity import SonosEntity
+from .speaker import SonosSpeaker
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_BATTERY_LEVEL = "battery_level"
+ATTR_BATTERY_CHARGING = "charging"
+ATTR_BATTERY_POWERSOURCE = "power_source"
+
+EVENT_CHARGING = {
+    "CHARGING": True,
+    "NOT_CHARGING": False,
+}
+
+
+def fetch_battery_info_or_none(soco: SoCo) -> dict[str, Any] | None:
+    """Fetch battery_info from the given SoCo object.
+
+    Returns None if the device doesn't support battery info
+    or if the device is offline.
+    """
+    with contextlib.suppress(ConnectionError, TimeoutError, SoCoException):
+        return soco.get_battery_info()
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Sonos from a config entry."""
+
+    sonos_data = hass.data[DATA_SONOS]
+
+    async def _async_create_entity(speaker: SonosSpeaker) -> SonosBatteryEntity | None:
+        if battery_info := await hass.async_add_executor_job(
+            fetch_battery_info_or_none, speaker.soco
+        ):
+            return SonosBatteryEntity(speaker, sonos_data, battery_info)
+        return None
+
+    async def _async_create_entities(speaker: SonosSpeaker):
+        if entity := await _async_create_entity(speaker):
+            async_add_entities([entity])
+        else:
+            async_dispatcher_send(
+                hass, f"{SONOS_ENTITY_CREATED}-{speaker.soco.uid}", SENSOR_DOMAIN
+            )
+
+    async_dispatcher_connect(hass, SONOS_DISCOVERY_UPDATE, _async_create_entities)
+
+
+class SonosBatteryEntity(SonosEntity, Entity):
+    """Representation of a Sonos Battery entity."""
+
+    def __init__(
+        self, speaker: SonosSpeaker, sonos_data: SonosData, battery_info: dict[str, Any]
+    ):
+        """Initialize a SonosBatteryEntity."""
+        super().__init__(speaker, sonos_data)
+        self._battery_info = battery_info
+
+    async def async_added_to_hass(self) -> None:
+        """Register polling callback when added to hass."""
+        await super().async_added_to_hass()
+
+        self.async_on_remove(
+            self.hass.helpers.event.async_track_time_interval(
+                self.update, BATTERY_SCAN_INTERVAL
+            )
+        )
```

Comment:
Do we still need this if events are working?

Maybe where the `_poll_timer` is created in `speaker` we could create a `_battery_poll_timer` if events aren't working and fire a similar signal

Reply:
I'm not certain how reliable the battery events are. Removing/attaching power triggers immediately, but unclear on battery drain updates. Running tests now to confirm.

Reply:
My tests show that events related to the battery are rare unless some other action triggers them. It's probably best to keep the (infrequent) polling timer if this is going to have use.

Reply:
Added a guard to defer polling with an executor if an event was handled recently.